### PR TITLE
[rust] dependency upgrades to fix cargo audit warnings

### PR DESCRIPTION
**Pull request by nbdd0121**
Fix (almostly) #18970

* Update `nix` to 0.26
	* requires some small code changes
* Update `serialport` to 4.2.1
	* requires patch update
* Switch from `atty` to `is-terminal`
	* requires minor code updates
* `cargo update` rest of dependencies

This addresses half of our cargo-audit warnings. The rest are:
* `time`, the affected version is still a dependency of chrono which mdbook uses, but the affected API is not used.
* `mach`, dependency of `serialport` but is only used on MacOS
* `ansi_term`, dependency of serde-annotate

Since `time` and `mach` does not affect us, I suppressed their warnings by adding relevant security advisory to `audit.toml` ignore section.

This lefts us with just `ansi_term`.

### CODE REVIEW

**Comment by nbdd0121** [comment]
Blocked until #19208 goes in.

**Comment by dmcardle** [comment]
Hi Gary, thanks for working on this!

Out of curiosity, how have you been running cargo audit?

File: third_party/rust/patches/serialport-pci-workaround.patch [hunk 1]
```diff
@@ -1,8 +1,8 @@
 diff --git a/src/posix/enumerate.rs b/src/posix/enumerate.rs
-index 65c42bb..ee50f55 100644
+index bec8f86...0000000 100644
```
Comment (dmcardle) [inline third_party/rust/patches/serialport-pci-workaround.patch:2]: FYI @lschuermann, now I remember you mentioned that this patch didn't apply cleanly. Thanks @nbdd0121 for updating it!

**Review by dmcardle — APPROVED** [review]
LGTM!

**Comment by nbdd0121** [comment]
The previous CI failure is caused by serde-derive starting to use precompiled binary but the file is not considered as data file in Rust rules so Bazel excludes them from sandbox. As a short-term solution I pinned Bazel to 1.0.171 which is the last version without precompiled binary,

File: third_party/rust/Cargo.toml [hunk 2]
```diff
@@ -55,8 +55,8 @@ raw_tty = "0.1.0"
 regex = "1.7"
 rsa = "0.8"
 rusb = "0.8.1"
-serde = { version="1.0", features=["serde_derive"] }
-serde_derive = "1.0"
+# serde-derive 1.0.172 uses precompiled binary which does not work well with Bazel.
+serde = { version="=1.0.171", features = ["derive"] }
```
Comment (cfrantz) [inline third_party/rust/Cargo.toml:59]: Can you create an issue in lowRISC/opentitan for this and leave a TODO comment here:
```
# TODO(opentitan#xxxxx): serde-derive 1.0.172 uses ...
```

**Review by cfrantz — APPROVED** [review]

**Review by HU90m — APPROVED** [review]

**Pull request merged**